rom/dev/src/flow/debug_unlock.rs

```diff
@@ -0,0 +1,315 @@
+            CfiCounter::delay();
+
+            match mbox.peek_recv() {
+                Some(txn) => break txn,
+                None => continue,
+            }
+        };
+
+        if CommandId::from(txn.cmd()) != CommandId::PRODUCTION_AUTH_DEBUG_UNLOCK_REQ {
+            Err(CaliptraError::ROM_SS_DBG_UNLOCK_PROD_INVALID_REQ_MBOX_CMD)?
+        }
+
+        let mut txn = txn.start_txn();
+        let mut request = ProductionAuthDebugUnlockReq::default();
+        FirmwareProcessor::copy_req_verify_chksum(&mut txn, request.as_bytes_mut())?;
+
+        let payload_length = |length: [u8; 3]| {
+            let mut len: usize = 0;
+            len |= length[0] as usize;
+            len |= (length[1] as usize) << 8;
+            len |= (length[2] as usize) << 16;
+            len * size_of::<u32>()
+        };
+
+        // Validate payload
+        if payload_length(request.length)
+            != size_of::<ProductionAuthDebugUnlockReq>() - size_of::<MailboxReqHeader>()
+        {
+            Err(CaliptraError::ROM_SS_DBG_UNLOCK_PROD_INVALID_REQ)?
+        }
+        // [TODO][CAP2] what do these 3 bytes mean when only 4 bits are active?
+        // Debug level
+        let dbg_level = payload_length(request.unlock_category);
+        if dbg_level & 0xf != dbg_level {
+            Err(CaliptraError::ROM_SS_DBG_UNLOCK_PROD_INVALID_REQ)?
+        }
+
+        let length = (size_of::<ProductionAuthDebugUnlockChallenge>()
+            - size_of::<MailboxReqHeader>())
+            / size_of::<u32>();
+        let challenge = env.trng.generate()?.as_bytes().try_into().unwrap();
+        let challenge_resp = ProductionAuthDebugUnlockChallenge {
+            vendor_id: request.vendor_id,
+            object_data_type: request.object_data_type,
+            length: length.to_ne_bytes()[..3].try_into().unwrap(),
+            unique_device_identifier: {
+                let mut id = [0u8; 32];
+                id[..17].copy_from_slice(&env.soc_ifc.fuse_bank().ueid());
+                id
+            },
+            challenge,
+            ..Default::default()
+        };
+        txn.send_response(challenge_resp.as_bytes())?;
+        core::mem::drop(txn);
+
+        // AUTH_DEBUG_UNLOCK_TOKEN
+        let txn = loop {
+            // Random delay for CFI glitch protection.
+            CfiCounter::delay();
+
+            match mbox.peek_recv() {
+                Some(txn) => break txn,
+                None => continue,
+            }
+        };
+
+        if CommandId::from(txn.cmd()) != CommandId::PRODUCTION_AUTH_DEBUG_UNLOCK_TOKEN {
+            Err(CaliptraError::ROM_SS_DBG_UNLOCK_PROD_INVALID_TOKEN_MBOX_CMD)?
+        }
+
+        env.soc_ifc.set_ss_dbg_unlock_in_progress(true);
+
+        let mut txn = txn.start_txn();
+        let mut request = ProductionAuthDebugUnlockToken::default();
+        FirmwareProcessor::copy_req_verify_chksum(&mut txn, request.as_bytes_mut())?;
+
+        let result: CaliptraResult<()> = (|| {
```

Comment:
The `handle_production()` function is getting long enough that maybe it is worth splitting this part out into a separate function.